File: Clase semana 06/02 - Clases grabadas/2023/1C/Curso.cpp
/*
 * Curso.cpp
 *
 *  Created on: 21/04/2023
 *      Author: algo2
 */

#include "Curso.h"
#include "Estudiante.h"

Curso::Curso(std::string codigo, std::string nombre, unsigned int cupoMaximo) {
	if (codigo == "") {
		throw "El codigo no puede ser vacio";
	}
	if (nombre == "") {
		throw "El nombre no puede ser vacio";
	}
	if (cupoMaximo < 1) {
		throw "El cupo maximo no puede ser 0";
	}
	this->cupoMaximo = cupoMaximo;
	this->codigo = codigo;
	this->nombre = nombre;
	this->inscriptos = new Estudiante*[this->cupoMaximo];
	for(int i = 0; i < this->cupoMaximo; i++) {
		this->inscriptos[i] = NULL;
	}
}

Curso::~Curso() {
	delete [] this->inscriptos;
}

std::string Curso::getNombre() {
	return this->nombre;
}

std::string Curso::getCodigo() {
	return this->codigo;
}

unsigned int Curso::getCupoMaximo() {
	return this->cupoMaximo;
}


Estudiante * Curso::getInscripto(unsigned int numero) {
	validarPosicion(numero);
	return this->inscriptos[numero - 1];
}

int Curso::inscribir(Estudiante * estudiante) {
	if (estudiante == NULL) {
		throw "EL estudiante es invalido";
	}
	this->inscriptos[getCantidadDeInscriptos()] = estudiante;
	return getCantidadDeInscriptos();
}

void Curso::desinscribir(unsigned int numero) {
	validarPosicion(numero);
	this->inscriptos[numero - 1] = NULL;
}

unsigned int Curso::getCantidadDeInscriptos() {
	for(int i = 0; i < this->cupoMaximo; i++) {
		if (this->inscriptos[i] == NULL) {
			return i;
		}
	}
	return this->cupoMaximo;
}

void Curso::validarPosicion(int numero) {
	if (numero < 1) {
		throw "No puede ser menor a 1";
	}
	if (numero > getCantidadDeInscriptos()) {
		throw "No puede ser mayor a cantidad de inscriptos";
	}
}

File: Clase semana 06/02 - Clases grabadas/2024/2C/src/ar/uba/fi/cb100/semana03/jueves/g3/n2/p2/Curso.java
package ar.uba.fi.cb100.semana03.jueves.g3.n2.p2;

public class Curso {

	private String nombre = null;
	private int cupo = 0;
	private Estudiante[] estudiantes = null;
	
	/**
	 * pre:
	 * @param nombre: no puede ser vacio
	 * @param cupo: debe ser mayor a 0
	 * @throws Exception: si los parametros no validan, tira excepcion
	 * 
	 * crea un curso con los parametros dados y la cantidad maxima de estudiante en cupo
	 */
	public Curso(String nombre, int cupo) throws Exception {
		this.setNombre(nombre);
		this.setCupo(cupo);
		this.estudiantes = new Estudiante[cupo];
		for(int i = 0; i < this.estudiantes.length; i++) {
			this.estudiantes[i] = null;
		}
	}
	
	/**
	 * pre:
	 * @param estudiante: no puede ser vacio
	 * @throws Exception valida el estudiante que no sea vacio, que no este ya inscripto, que tenga cursos libres
	 *                   y que el curso tenga cupo
	 */
	public void inscribir(Estudiante estudiante) throws Exception {
		if (estudiante == null) {
			throw new Exception("El estudiante no puede ser vacio");
		}
		for(int i = 0; i < this.estudiantes.length; i++) {			
			if (this.estudiantes[i] == null) {
				if (estudiante.getCantidadDeCursosLibre() <= 0) {
					throw new Exception("El estudiante ya no tiene cursos libres para ser inscripto");
				}
				this.estudiantes[i] = estudiante;
				estudiante.inscribir(this);
			} else {
				if (this.estudiantes[i].esIgual(estudiante)) {
					throw new Exception("El estudiante ya esta inscripto");
				}
			}
		}
		throw new Exception("No hay cupo");
	}
	
	/**
	 * 
	 * @return
	 */
	public String getNombre() {
		return nombre;
	}
	
	/**
	 * 
	 * @return
	 */
	public int getCupo() {
		return cupo;
	}
	
	/**
	 * 
	 * @return
	 */
	public Estudiante[] getEstudiantes() {
		return estudiantes;
	}
	
	/**
	 * 
	 * @param nombre
	 * @throws Exception 
	 */
	public void setNombre(String nombre) throws Exception {
		if (Herramientas.estaVacio(nombre)) {
			throw new Exception("El nombre no puede ser vacio");
		}
		this.nombre = nombre;
	}
	
	/**
	 * 
	 * @param cupo
	 * @throws Exception 
	 */
	public void setCupo(int cupo) throws Exception {
		if (!Herramientas.esPositivo(cupo)) {
			throw new Exception("No puede ser menor a 1");
		}
		if (cupo <= this.estudiantes.length) {
			throw new Exception("No puede ser menor a la cantidad de estudiantes " + this.estudiantes.length);
		}
		Estudiante[] estudiantesTemporal = new Estudiante[cupo];
		for(int i = 0; i < cupo; i++) {
			estudiantesTemporal[i] = this.estudiantes[i];
		}
		this.estudiantes = estudiantesTemporal;
		this.cupo = cupo;
	}
	
	/**
	 * 
	 * @param estudiantes
	 */
	public void setEstudiantes(Estudiante[] estudiantes) {
		this.estudiantes = estudiantes;
	}
	
	
	
	
}
